Waterlines

1ec5 opened the pull request:
Bodies of water are filled in a pastel blue and are outlined with a medium blue hairline. The outline is hidden at low zoom levels except along coastlines. Intermittent lakes have a dashed outline. Sea and ocean labels have a halo that matches the fill color for bodies of water, which will result in a knockout effect more often than a halo. Park boundaries along coastlines are blended with the waterline; park boundaries over water are much more subtle, no longer appearing to glow.

Before | After
----|----
[<img src="https://user-images.githubusercontent.com/1231218/206672352-bf29f1d4-bc83-4190-9f4e-337a2c039baa.png" width="400" alt="Gulf of Mexico before">](https://zelonewolf.github.io/openstreetmap-americana/#map=4/28.68/-92.42) | <img src="https://user-images.githubusercontent.com/1231218/206672383-99b5e13b-3f81-4c3c-8af2-25f46e682a4d.png" width="400" alt="Gulf of Mexico after">
[<img src="https://user-images.githubusercontent.com/1231218/206673205-52d2b008-baf5-46da-84c9-92bb9cc21410.png" width="400" alt="Southeast Louisiana before">](https://zelonewolf.github.io/openstreetmap-americana/#map=7/29.623/-89.629) | <img src="https://user-images.githubusercontent.com/1231218/206673234-3cfab3d2-08af-4312-80ea-f3a720c81555.png" width="400" alt="Southeast Louisiana after">
[<img src="https://user-images.githubusercontent.com/1231218/206673802-6e09796c-9237-4af0-b8af-a89d7303a2e0.png" width="400" alt="Mississippi Delta before">](https://zelonewolf.github.io/openstreetmap-americana/#map=8/29.184/-89.459) | <img src="https://user-images.githubusercontent.com/1231218/206673817-0d73a570-86f3-4f62-848c-7e05f8c57aa4.png" width="400" alt="Mississippi Delta after">
[<img src="https://user-images.githubusercontent.com/1231218/206674505-bb37774b-35d5-4677-9061-ccc84f729a01.png" width="400" alt="Venice before">](https://zelonewolf.github.io/openstreetmap-americana/#map=12/29.27786/-89.34858) | <img src="https://user-images.githubusercontent.com/1231218/206674530-414382a1-5f03-490b-9d84-7e5d504f548b.png" width="400" alt="Venice after">
[<img src="https://user-images.githubusercontent.com/1231218/206675082-5a752144-0258-4235-ad32-a25b67fecd43.png" width="400" alt="Mobile Bay before">](https://zelonewolf.github.io/openstreetmap-americana/#map=7/30.374/-87.621) | <img src="https://user-images.githubusercontent.com/1231218/206675103-2e1998bb-551b-4e09-97ba-84668cbf6718.png" width="400" alt="Mobile Bay after">
[<img src="https://user-images.githubusercontent.com/1231218/206675622-72979e8e-8f72-499b-9889-36e5ae540c1b.png" width="400" alt="Mobile before">](https://zelonewolf.github.io/openstreetmap-americana/#map=8/30.45/-87.893) | <img src="https://user-images.githubusercontent.com/1231218/206675639-4207b5e4-7a25-4229-a5f3-28bc6c0b32b4.png" width="400" alt="Mobile after">
[<img src="https://user-images.githubusercontent.com/1231218/206676399-92ab3ab0-48b2-4079-a9cc-5cc068eb47e9.png" width="400" alt="Pensacola before">](https://zelonewolf.github.io/openstreetmap-americana/#map=10/30.3448/-87.3757) | <img src="https://user-images.githubusercontent.com/1231218/206676428-2eee28f6-1d67-45f9-b872-a1a80534ad3a.png" width="400" alt="Pensacola after">
[<img src="https://user-images.githubusercontent.com/1231218/206677081-2c773118-7389-43d1-a0a7-6f7a01e305b9.png" width="400" alt="Florida Keys before">](https://zelonewolf.github.io/openstreetmap-americana/#map=5/25.17/-79.2) | <img src="https://user-images.githubusercontent.com/1231218/206677095-6b287baf-d0cb-4e36-8729-65f640de5b77.png" width="400" alt="Florida Keys after">
[<img src="https://user-images.githubusercontent.com/1231218/206677644-a3b1e1e3-1736-4a75-a52b-988a7db88bcb.png" width="400" alt="Gulf of California before">](https://zelonewolf.github.io/openstreetmap-americana/#map=4/29.09/-113.27) | <img src="https://user-images.githubusercontent.com/1231218/206677676-58c723ca-1a79-43d7-9918-15b01af02b23.png" width="400" alt="Gulf of California after">
[<img src="https://user-images.githubusercontent.com/1231218/206678376-a24ffb6a-232b-4f74-835e-f25fc53880f8.png" width="400" alt="New York Harbor before">](https://zelonewolf.github.io/openstreetmap-americana/#map=12/40.69992/-74.01363) | <img src="https://user-images.githubusercontent.com/1231218/206678392-7d176ab0-ffc6-4205-92ae-93f45a6c4ccc.png" width="400" alt="New York Harbor after">
[<img src="https://user-images.githubusercontent.com/1231218/206679494-724438d4-0f98-4357-927a-27c3f9592b2a.png" width="400" alt="Grand Lake St. Marys before">](https://zelonewolf.github.io/openstreetmap-americana/#map=12/40.51248/-84.44934) | <img src="https://user-images.githubusercontent.com/1231218/206679513-1ad2a363-8627-47b6-8e87-84767a487996.png" width="400" alt="Grand Lake St. Marys after">
[<img src="https://user-images.githubusercontent.com/1231218/206680127-3c2e3aa8-28e0-47a4-a8fd-867daab2a08f.png" width="400" alt="Fairfield before">](https://zelonewolf.github.io/openstreetmap-americana/#map=13/39.33883/-84.59982) | <img src="https://user-images.githubusercontent.com/1231218/206680138-3fe3aa26-9cab-48bb-a34e-a76215dc5c60.png" width="400" alt="Fairfield after">
[<img src="https://user-images.githubusercontent.com/1231218/206671437-6adb4a03-80c9-42d1-bc45-5aba58db0453.png" width="400" alt="Kennecott before">](https://zelonewolf.github.io/openstreetmap-americana/#map=11/40.7521/-112.1435) | <img src="https://user-images.githubusercontent.com/1231218/206671411-6862d931-68f4-41e2-9ed8-e0352da9d610.png" width="400" alt="Kennecott after">
[<img src="https://user-images.githubusercontent.com/1231218/206681615-b7fd009c-a497-49a2-b3e7-2b74c85dd52f.png" width="400" alt="Alviso before">](https://zelonewolf.github.io/openstreetmap-americana/#map=13/37.4356/-121.98692) | <img src="https://user-images.githubusercontent.com/1231218/206681632-db17031a-8e5c-440b-8104-d7bfcc2f36db.png" width="400" alt="Alviso after">

Fixes #608.

/ref gravitystorm/openstreetmap-carto#3468

1ec5 commented on src/americana.js:
```diff
@@ -45,9 +45,11 @@ function buildLayers() {
     lyrWater.waterwayIntermittent,
+    lyrWater.waterLine,
+    lyrWater.waterLineIntermittent,
+    lyrWater.water,
```
The water fill layer needs to draw over the waterway line layer; otherwise, there will be a prominent line down the middle of the river that looks like a seam.

1ec5 commented on src/americana.js:
```diff
@@ -45,9 +45,11 @@ function buildLayers() {
-    lyrWater.water,
     lyrWater.waterway,
     lyrWater.waterwayIntermittent,
+    lyrWater.waterLine,
```
The waterline layer needs to draw under the water fill layer; otherwise, there would be visible seams at tile boundaries over water and at the mouths of rivers and harbors.

1ec5 replied:
Although there isn’t a seam at mouths of rivers, @zekefarwell pointed out that the waterline ends abruptly at the point where the `water=river` begins, seen here at zoom level 7.99 (but not at z7 or z8):

[<img src="https://user-images.githubusercontent.com/1231218/206877068-f02e4e00-3f83-42c8-9881-31ba641315a4.png" width="400" alt="St. Lawrence">](https://zelonewolf.github.io/openstreetmap-americana/#map=7.99/46.999/-70.696)

My initial inclination was not to worry much about it. Paper maps are unapologetic about beginning and ending strokes where they stop being applicable.

I tried giving lakes and rivers a fainter waterline at low zoom levels, but any color that’s discernible from the water fill ends up overemphasizing the water feature:

[<img src="https://user-images.githubusercontent.com/1231218/206877474-5e1fb447-e883-43ca-9826-8796cbb9f60d.png" width="400" alt="Dallas no waterline">](https://zelonewolf.github.io/openstreetmap-americana/#map=6/32.777/-96.655) <img src="https://user-images.githubusercontent.com/1231218/206877477-9fc3f025-3c76-44e3-9b93-58c4b944e389.png" width="400" alt="Dallas waterline">

zekefarwell replied:
I tried out some variations for inland water as well.  It's definitely a challenge.  Between z6 and z8, there are many visible water bodies with quite highly detailed geometry.  The waterline can look ok on smoother water bodies, but on complex shorelines (reservoirs for example) the line ends zig-zagging back and for so much that it ends up looking thicker and/or darker.  I found it can look quite good below z6 as the water bodies become much more generalized at that point and many smaller ones are no longer visible.  I'm optimistic we can find a way to make a water line of some sort work on inland water bodies, but I'm happy to leave that for a future PR.

1ec5 commented on src/layer/water.js:
```diff
@@ -64,13 +64,66 @@ export const water = {
       0.3,
       1,
     ],
+    "fill-outline-color": Color.waterFillTranslucent,
```
Because the waterline is under the water fill layer, only the half of the line that lies outside the waterbody would normally be visible. This outline doesn’t apply the waterline color. By applying some extra blue, it seems to help the waterline overcome the fill’s antialiasing and show through more prominently than it would on its own. The waterline relies on color instead of width because any additional width makes some waterways look cartoonish.

jleedev replied:
Try experimenting on waterLine with (say) `{"line-width":1.25,"line-blur":0.25}`, especially at 1x vs 2x. Could smooth out jaggies and make it more subtle, or just look terrible.

1ec5 replied:
The `line-width` makes the line darker (not wider), but the addition of a `line-blur` seems to cancel out the change on a 2× display. I don’t see any difference from the original state of this branch. Is there a particular area I should be testing?

jleedev replied:
Not sure, just an idea. The lakes in eastern Ohio are pretty noisy though.

1ec5 replied:
Sorry, I’m not seeing the `line-width` and `line-blur` make any difference there either. If you mean that the waterlines make the reservoirs in eastern Ohio look much more prominent, I’m seeing the same thing in North Texas, especially when introducing waterlines below z8: https://github.com/ZeLonewolf/openstreetmap-americana/pull/610#discussion_r1045143790.

1ec5 commented on src/layer/water.js:
```diff
@@ -64,13 +64,66 @@ export const water = {
+    [">=", ["zoom"], 8],
+    mediumRivers,
+    [">=", ["zoom"], 16],
+    "lake",
+    ["all", ["!=", ["get", "intermittent"], 1], [">=", ["zoom"], 8]],
```
Unfortunately, there’s no way to distinguish between a small lake and a large one. The Great Lakes conspicuously lack waterlines until zoom level 8. But for smaller lakes near larger bodies of water, such as Lake Okeechobee, this lack of a waterline looks more intentional.

jleedev replied:
It might just be from years of staring at a lack of Great Lakes in the osm coastline, but it looks intentional to me.

The eastern seaboard looks a little worse.

![coastline1](https://user-images.githubusercontent.com/23022/206701915-0e8d1683-37e6-4ab9-9d0b-3b8d29939ad8.png)
![coastline2](https://user-images.githubusercontent.com/23022/206701919-96fe1c2a-f601-43cc-af59-a4481f387267.png)
![coastline3](https://user-images.githubusercontent.com/23022/206701921-a5d08b15-e9d2-486e-8305-371547ed14f1.png)

1ec5 replied:
Are you referring to the North Carolina coastline? I’m unsure why none of the waters behind the barrier islands are `ocean` in the vector tiles; it’s inconsistent with similar waters elsewhere along the Eastern Seaboard and Gulf Coast.

ZeLonewolf replied:
It's no mystery at all: https://www.openstreetmap.org/way/815122837

Someone pulled a Chesapeake Bay and put the coastline outside.

ZeLonewolf replied:
Also, Virginia's maritime boundary is missing there?

1ec5 replied:
In #35, we attempted to hide maritime boundaries except for boundaries between two administrative areas. The vector tiles have North Carolina’s boundary over the Atlantic as `maritime: 0`, whereas it’s `maritime: 1` for the other surrounding states.

ZeLonewolf replied:
Ah, I see.  Thanks @Mashin6 for fixing this a couple days ago :-D

1ec5 replied:
> It might just be from years of staring at a lack of Great Lakes in the osm coastline, but it looks intentional to me.

Yes, at zoom level 3, there’s nothing unseemly about the shoreline being absent from the Great Lakes as with any other lake. However, it does start to feel like something is missing at z6 or so, when the coastline isn’t likely to be within view anymore. I’m of two minds about this discrepancy: either we’re OK with it on the basis that the waterline at low zoom levels only reflects `natural=coastline` in OSM, or it’s annoying and we should come up with a solution at the data level for inland seas.

ZeLonewolf replied:
I've been slowly fixing the NC Outer Banks' coastline to be where it's supposed to be.  This style change does a great job of highlighting the coastline is in the wrong place, and I consider that a feature and not a bug.

1ec5 commented:
The screenshots above are from a 2× display. Since the waterline is such a delicate hairline, we should make sure it’s still legible on a 1× display.

jleedev commented:
The main issue this seems to highlight is not mapping but data processing, short segments of river and riverbank disappearing at low zooms and creating gaps.

![data1](https://user-images.githubusercontent.com/23022/206702243-8308b017-9b9f-4b12-8ceb-5c0ee2cab1ef.png)
![data2](https://user-images.githubusercontent.com/23022/206702244-ef8b5c50-9c43-498e-8e9f-a3ae6b6e7903.png)

jleedev commented:
> The screenshots above are from a 2× display. Since the waterline is such a delicate hairline, we should make sure it’s still legible on a 1× display.

This spot gave me pause:

![Screenshot 2022-12-09 07 10 42](https://user-images.githubusercontent.com/23022/206702320-cdb80885-f33a-4dc9-89c2-d2e4fb66de3a.png)

Perhaps: `"line-opacity": ["interpolate",["exponential",1.25],["zoom"],7,0,10,1]`.

1ec5 commented:
> The main issue this seems to highlight is not mapping but data processing, short segments of river and riverbank disappearing at low zooms and creating gaps.

These gaps already exist on the main branch, but the waterline accentuates them as if they’re enormous oxbow lakes. For what it’s worth, I’ve been mulling a change to hide all `river` water areas at mid zoom levels, like most maps do, but it we’d need to experiment with various cutoffs. I was thinking that could be tail work.

zekefarwell commented:
> The screenshots above are from a 2× display. Since the waterline is such a delicate hairline, we should make sure it’s still legible on a 1× display.

Just tested on a 1x display.  The line ends up blurring, making it a bit thicker and a bit ligher than on a 2x display, but it looks fine.
![cape-code-water-outline-1x png](https://user-images.githubusercontent.com/281482/206756452-afcc1fc9-b8d1-4d23-80ab-14e590ad6872.png)

From a 2x display for comparison.
<img width="676" alt="cape-code-water-outline-2x" src="https://user-images.githubusercontent.com/281482/206756475-3f9f9f34-3487-4235-a8be-4113c85a5516.png">

ZeLonewolf commented:
Overall I'm pretty happy with this and I'd be good with merging in its current state as an improvement over what we have presently.

I'd like to discuss adding the line treatment to inland lakes also but that can be a separate discussion.

Also the line treatment looks a little funny when the coastline is very complex, but I don't think that's a showstopper either:

![image](https://user-images.githubusercontent.com/3254090/206884603-8a8b2528-7bb7-4001-9a4f-ab0ab8cf8385.png)

Are there any open issues that we think need to be hashed out?

zekefarwell reviewed (approved):
This is looking really good to me.  At first I wasn't quite sure about the water fill.  It seemed maybe a bit too light.  After looking at it for a while now, I think it's just about perfect.  The hairline technique you've come up with here certainly works much better than what I was trying out last year.  

I also would like to see some sort of waterline on inland water bodies at all zoom levels, but it's definitely a challenge and can be looked into later.  Some other things to note for potential future improvement:
- Some admin boundaries seems a bit too dark now that the water is much lighter.  
- Rivers and streams mapped only as lines use the water outline color now and this looks good when they are thin.  As the zoom increases and they get thicker, it starts to look off though.  Maybe river lines need casing like roads at a high enough zoom?
<img width="118" alt="Screen Shot 2022-12-11 at 1 37 36 AM" src="https://user-images.githubusercontent.com/281482/206890063-e95fc63b-b9f5-42c3-a013-41bdf5d5331c.png">
<img width="583" alt="Screen Shot 2022-12-11 at 1 37 50 AM" src="https://user-images.githubusercontent.com/281482/206890067-59fd204a-5e45-405f-8a34-b9d3020742a1.png">

1ec5 commented:
> This is looking really good to me. At first I wasn't quite sure about the water fill. It seemed maybe a bit too light. After looking at it for a while now, I think it's just about perfect. The hairline technique you've come up with here certainly works much better than what I was trying out last year.

Glad to hear it. Changing the water color really affects the style’s overall mood, so I don’t take this choice lightly. I went back and forth between this and a darker shade for a while, but this lighter shade has grown on me over the past few days, probably because of the additional contrast with the waterline and labels. Plus, we save a little on ink.

> Rivers and streams mapped only as lines use the water outline color now and this looks good when they are thin. As the zoom increases and they get thicker, it starts to look off though. Maybe river lines need casing like roads at a high enough zoom?

Looks like this will happen at z14–15 and especially z16+ any time a `waterway=river` that hasn’t been mapped as an area meets one that has, as with the [Browns River](https://www.openstreetmap.org/way/539727654) meeting the [Lamoille River](https://www.openstreetmap.org/way/366824665). For a river as narrow as the Browns River, it might be misleading to give the line a casing as though it has been mapped as an area and only takes up that small amount of space. I’m also unsure how to case the dashed line we’re using for intermittent waterways.

Instead, I’ve changed the waterway layer to fade to the water fill color by z14 and moved the water line layer under the waterway layer:

[<img src="https://user-images.githubusercontent.com/1231218/206897306-fbc773a3-7a2d-4778-8674-4c4454c2528d.png" width="400" alt="z13">](https://zelonewolf.github.io/openstreetmap-americana/#map=13/44.65965/-73.01807) [<img src="https://user-images.githubusercontent.com/1231218/206897361-10843e97-40d6-45d5-bd81-a3816941ff32.png" width="400" alt="z14">](https://zelonewolf.github.io/openstreetmap-americana/#map=14/44.66087/-73.0185) [<img src="https://user-images.githubusercontent.com/1231218/206897378-19907ca0-fe94-48a8-a980-ab4a26f4a9dd.png" width="400" alt="z15">](https://zelonewolf.github.io/openstreetmap-americana/#map=15/44.66087/-73.0185) [<img src="https://user-images.githubusercontent.com/1231218/206897424-ed03562e-8897-41b3-af0d-22a333fd4214.png" width="400" alt="z16">](https://zelonewolf.github.io/openstreetmap-americana/#map=16/44.660472/-73.018908)

I think this is an improvement because there’s less of a need for the waterline to provide definition at this zoom level. It also avoids a mixed message to mappers that the waterline sometimes represents the edge of an area they should map but sometimes represents a line casing.

With this latest change, the waterway layer will continue to show through an intermittent water area, such as where the banks of a river fluctuate seasonally. However, the overall appearance does feel different than before this PR. Before, the surrounding area looked like residual moisture from a previous rainy season, whereas now it looks much less skeuomorphic and more like something you’d see on a traditional paper map:

[<img src="https://user-images.githubusercontent.com/1231218/206897901-a9521654-9647-4850-be9b-fc9231f564c9.png" width="400" alt="Colorado before">](https://zelonewolf.github.io/openstreetmap-americana/#map=10/36.1231/-114.0735) <img src="https://user-images.githubusercontent.com/1231218/206897914-2a137f7e-aaac-4202-ae88-236d6bb94ca6.png" width="400" alt="Colorado after">

Unfortunately, the lighter color makes intermittent waterways quite subtle within parks. I’m not so concerned about the main channel of the Colorado here, because the whole thing is the Colorado, but the side streams are barely perceptible even without color blindness:

[<img src="https://user-images.githubusercontent.com/1231218/206898204-8fd62626-b1d4-4032-9b66-2fd864af6bac.png" width="400" alt="Colorado intermittent before">](https://zelonewolf.github.io/openstreetmap-americana/#map=14/36.12649/-114.09572) <img src="https://user-images.githubusercontent.com/1231218/206898227-7ad70282-2f4b-450b-8129-8bb93b1cdf91.png" width="400" alt="Colorado intermittent after">

In principle, we could put a land-colored knockout behind the waterway but in front of the park fill to provide the necessary contrast. Some print maps take a similar coloring book approach to land use fills. But a dashed water fill color cased in near-white surrounded by mint green comes out as a light gray.

[<img src="https://user-images.githubusercontent.com/1231218/206899434-e317864a-1c78-40db-9674-b7711ddc7228.png" width="400" alt="gray">](https://zelonewolf.github.io/openstreetmap-americana/#map=14/36.13308/-114.08804)

For now, I’m keeping intermittent waterways in the darker blue, matching waterlines. It would be counterintuitive for a legend, but at least the dashing reduces the user’s expectation that the lines will join up perfectly:

<img src="https://user-images.githubusercontent.com/1231218/206899591-e7261a24-3a17-45d0-8c89-abe9c5177d78.png" width="400" alt="darker">

1ec5 commented:
> Some admin boundaries seems a bit too dark now that the water is much lighter.

The boundary line might be difficult to lighten because it also needs to maintain contrast against the boundary casing. However, I lightened ferry lines and labels and turn the halo into a knockout to mitigate the density in some areas, now that it’s easier to maintain contrast against the water fill. Boundaries are slightly more discernible amid a sea of ferry lines:

[<img src="https://user-images.githubusercontent.com/1231218/206900595-4b80f264-883f-40cc-9c42-03db96594a26.png" width="400" alt="New York Harbor before">](https://zelonewolf.github.io/openstreetmap-americana/#map=12/40.70042/-74.0194) <img src="https://user-images.githubusercontent.com/1231218/206900634-86fff664-2bcf-4aca-9740-444ee73dc513.png" width="400" alt="New York Harbor after">

1ec5 commented:
Should we revisit the translucent place label halos in https://github.com/ZeLonewolf/openstreetmap-americana/pull/606#pullrequestreview-1211089504, now that it’s easier to achieve a knockout effect without implying a deformed coastline?

ZeLonewolf commented on scripts/taginfo_template.json:
```diff
@@ -298,6 +298,20 @@
+      "object_types": ["way"],
+      "description": "Coastlines are represented by a hairline at all zoom levels, except across the mouth of a river.",
+      "doc_url": "https://openmaptiles.org/schema/#water"
+    },
```
I don't believe that we actually consume OSM coastline tagging, but I might be wrong on that.

jleedev replied:
I believe it's natural earth <z6 and water-polygons ≥z6, the latter being an indirect use of coastline.

ZeLonewolf replied:
:+1: that works for me.
